homeassistant/components/bangolufsen/config_flow.py

```diff
@@ -0,0 +1,203 @@
+                return self.async_abort(reason="value_error")
+
+            self._client = MozartClient(self._host, urllib3_logging_level=logging.ERROR)
+
+            # Try to get information from Beolink self method.
+            try:
+                beolink_self = cast(
+                    ApplyResult[BeolinkPeer],
+                    self._client.get_beolink_self(async_req=True, _request_timeout=3),
+                ).get()
+
+            except (
+                ApiException,
+                NewConnectionError,
+                MaxRetryError,
+                NotFoundException,
+            ) as error:
+                return self.async_abort(
+                    reason={
+                        ApiException: "api_exception",
+                        NewConnectionError: "new_connection_error",
+                        MaxRetryError: "max_retry_error",
+                        NotFoundException: "not_found_exception",
+                    }[type(error)]
+                )
+
+            self._beolink_jid = beolink_self.jid
+            self._serial_number = beolink_self.jid.split(".")[2].split("@")[0]
+
+            await self.async_set_unique_id(self._serial_number)
+            self._abort_if_unique_id_configured()
+
+            return await self.async_step_confirm()
+
+        data_schema = {
+            vol.Required(CONF_HOST): str,
+            vol.Required(CONF_MODEL, default=DEFAULT_MODEL): SelectSelector(
+                SelectSelectorConfig(options=COMPATIBLE_MODELS)
+            ),
+        }
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(data_schema),
+        )
+
+    async def async_step_zeroconf(
+        self, discovery_info: ZeroconfServiceInfo
+    ) -> FlowResult:
+        """Handle discovery using Zeroconf."""
+
+        # Check if the discovered device is a Mozart device
+        if ATTR_FRIENDLY_NAME not in discovery_info.properties:
+            return self.async_abort(reason="not_mozart_device")
+
+        self._host = discovery_info.host
+        self._model = discovery_info.hostname[:-16].replace("-", " ")
+        self._serial_number = discovery_info.properties[ATTR_SERIAL_NUMBER]
+        self._beolink_jid = f"{discovery_info.properties[ATTR_TYPE_NUMBER]}.{discovery_info.properties[ATTR_ITEM_NUMBER]}.{self._serial_number}@products.bang-olufsen.com"
+
+        self.context["title_placeholders"] = {
+            "name": discovery_info.properties[ATTR_FRIENDLY_NAME]
+        }
+
+        await self.async_set_unique_id(self._serial_number)
+        self._abort_if_unique_id_configured()
+
+        return await self.async_step_confirm()
+
+    async def async_step_confirm(
+        self, user_input: UserInput | None = None
+    ) -> FlowResult:
+        """Confirm the configuration of the device."""
+        if user_input is not None:
+            data = await self._compile_data()
+            return self.async_create_entry(title=self._name, data=data)
```

Comment:
```suggestion
        if user_input is not None:
            self._client = MozartClient(self._host, urllib3_logging_level=logging.ERROR)
            data = await self._compile_data()
            return self.async_create_entry(title=self._name, data=data)
```

Reply:
I've created the client in the zeroconf method, to avoid the client being created twice in a manual config flow.

Reply:
Right. To avoid wasting memory if the user is not interested in adding their B&O speaker to Home Assistant, maybe add an intermediate step, `async_step_zeroconf_confirm` which creates the client and then calls `async_step_confirm`, or check if the client exists and create it if it doesn't in `_compile_data`.